velox/vector/BaseVector.cpp

```diff
@@ -539,7 +539,7 @@ void BaseVector::ensureWritable(
     // TODO Figure out how to allow memory reuse for a newly loaded vector.
     // LazyVector holds a reference to loaded vector, hence, unique() check
     // below will never pass.
-    VELOX_NYI();
+    result = BaseVector::loadedVectorShared(result);
```

Comment:
I get the impression that this should be handled at calling site:
1. If it is already loaded, loaded vector should be passed in and the reference to lazy should be released before that to allow reusing of the loaded vector
2. If it is not loaded, just allocate new vector

Reply:
What I have noticed is that its extremely difficult to ensure that we can get rid of the lazy reference in all codepaths:

> 1. If it is already loaded, loaded vector should be passed in and the reference to lazy should be released before that to allow reusing of the loaded vector

- Lazy vectors can be nested inside encoding layers and a vector pointing to any of the inside layers can be passed around. This means that if some codepath receives a lazy vector and you load it then some other vector might have a ref to the lazy vector, so we cannot release the lazy ref

> 2. If it is not loaded, just allocate new vector

- For similar reasons as above, if its not loaded, you need to load the vector, keep the lazy vectors's ref to the loaded vector and finally, to ensureWritable, you would have to make a copy.

Ideally, we would like to avoid that extra copy, but to make sure things dont break, I added this part. For reference, this copy was already being done, see description in #6168 that talks about what happens if the vector is not preloaded during expression eval.

Reply:
If the lazy vector is referenced else where we can leave it as it is and create a new one for reuse.  Here we just need to make sure if the lazy is unique, we should reuse its loaded vector.

Reply:
@Yuhta Thanks for clarifying. Just to confirm, with the change above, if the lazy is unique, its only held by 'result' so once we reassign it to its loaded vector, it should get destroyed which would make the loaded vector unique and re-usable, right?

Reply:
Yes good catch, it will be the caller's responsibility to release the reference on lazy before this call if we really want to reuse the loaded vector.